state-chain/pallets/cf-cfe-event-emitter/src/lib.rs

```diff
@@ -0,0 +1,146 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod benchmarking;
+mod weights;
+
+pub use weights::PalletWeight;
+use weights::WeightInfo;
+
+use cf_chains::{
+	btc::BitcoinCrypto, dot::PolkadotCrypto, evm::EvmCrypto, Bitcoin, Ethereum, Polkadot,
+};
+use cf_primitives::{Ed25519PublicKey, Ipv6Addr, Port};
+use cf_traits::{CfeEventEmitterForChain, CfeEventEmitterForCrypto, CfeEventEmitterT, Chainflip};
+use frame_support::{
+	pallet_prelude::Hooks, storage::types::StorageMap, traits::StorageVersion, Twox64Concat,
+};
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use sp_std::vec::Vec;
+
+pub const PALLET_VERSION: StorageVersion = StorageVersion::new(1);
+
+/// How long to keep cfe events for (in SC blocks)
+const EVENT_LIFETIME: u32 = 20;
+
+pub type EventId = u64;
+
+pub type CfeEvent<T> = cfe_events::CfeEvent<<T as Chainflip>::ValidatorId>;
+pub type ThresholdSignatureRequest<T, C> =
+	cfe_events::ThresholdSignatureRequest<<T as Chainflip>::ValidatorId, C>;
+pub type KeyHandoverRequest<T, C> =
+	cfe_events::KeyHandoverRequest<<T as Chainflip>::ValidatorId, C>;
+pub type KeygenRequest<T> = cfe_events::KeygenRequest<<T as Chainflip>::ValidatorId>;
+pub type TxBroadcastRequest<T, C> =
+	cfe_events::TxBroadcastRequest<<T as Chainflip>::ValidatorId, C>;
+
+#[frame_support::pallet]
+pub mod pallet {
+
+	use cf_traits::Chainflip;
+
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::storage_version(PALLET_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	#[pallet::disable_frame_system_supertrait_check]
+	pub trait Config: Chainflip {
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_cfe_events)]
+	#[pallet::unbounded]
+	pub type CfeEvents<T: Config> =
+		StorageMap<_, Twox64Concat, BlockNumberFor<T>, Vec<CfeEvent<T>>>;
```

Comment:
Why do we need a block number here? Seems like an unnecessary addition to me. When you query for the events, you need to provide a block hash that those events were stored at anyway. Having the block number here is misleading since, I immediately think "I can query events by block number" but that's only the case if I also query within that block number's hash and that hash + 20... is just much easier if we query by hash and hash only.

Reply:
> When you query for the events, you need to provide a block hash that those events were stored at anyway. 

This is only true if we immeidately clear the storage at the beginning of every new block (otherwise we'd also be getting events from previous blocks). Is that what you are suggesting? I wanted to give the CFE a bit of a buffer (EVENT_LIFETIME in this case) in case it falls behind and needs to catch up by reading older blocks. I'm worried that if we do what you are suggesting, then it would be possible for SC to clear events before CFE had a chance to read them (imagine a case where our node had a network hiccup and then received two blocks at once when recovered).

Reply:
Took me some time, but I see now what you are saying. Even if we overwrite the events and the CFE didn't get a chance to read them, they are still going to be available for some time when requested by the right block hash. Will simplify.

Reply:
Done. Note that I had to move `CfeEventEmitter` in Rutnime before any other pallet that may emit a cfe event so that we can delete old events in `on_initialize()` before any other event is added for the new block.